### PR TITLE
Dreambooth: reduce VRAM usage

### PR DESCRIPTION
**Pull request by gleb-akhmerov**
This allowed me to run Dreambooth on a 3060 with VRAM usage of 11124MiB using the following parameters:
```bash
--mixed_precision="fp16" \
--use_8bit_adam \
--enable_xformers_memory_efficient_attention \
--gradient_accumulation_steps=1 \
--gradient_checkpointing \
--resolution=512 \
--train_batch_size=1 \
--train_text_encoder
```

Also, I was able to use batches of size up to 5 images, with VRAM usage of 11694MiB.

`--resolution=768` is possible too, with VRAM usage of 11450MiB.
`--resolution=1024` uses 11598MiB, but it's pretty slow (about 4s/it).

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
cc @patil-suraj

**Review by patil-suraj — COMMENTED** [review]
Thanks a lot for the PR! It saves quite a bit of memory. But according to torch [docs](https://pytorch.org/docs/stable/generated/torch.optim.Optimizer.zero_grad.html), it changes certain behaviour. So I think instead of setting it to `True` by default we could make it a argument that the user can control, something like `--optimizer_set_to_none`, which will be `false` by default. Also cc @pcuenca @patrickvonplaten

**Comment by patrickvonplaten** [comment]
Agree with @patil-suraj here, @gleb-akhmerov could we maybe add an arguments `--set_grads_to_none` ? (maybe a bit better than `optimizer_set_to_none`

**Comment by gleb-akhmerov** [comment]
@patil-suraj @patrickvonplaten Good point! I've changed the code to include the option. Also, I've updated the readme to allow people to find this option more easily. What do you think?

**Comment by gleb-akhmerov** [comment]
By the way, there's a note in the [Fine-tune text encoder with the UNet section](https://github.com/huggingface/diffusers/blob/3ecbbd628825712552a1834f78a224bfab4cfe21/examples/dreambooth/README.md#fine-tune-text-encoder-with-the-unet):

> Note: Training text encoder requires more memory, with this option the training won't fit on 16GB GPU. It needs at least 24GB VRAM.

This is no longer the case if one uses the gradient checkpointing, 8-bit optimizer, xformers and the new optimization from this PR. However, I'm not sure how to change this note to reflect the current state of affairs. Could you help me with it?

**Review by patil-suraj — APPROVED** [review]
Looks great, thanks!
